File: ios/Classes/FaceScanner/FlutterFaceScanner.swift
//
//  FlutterFaceScanner.swift
//  ekyc_id_flutter
//
//  Created by Socret Lee on 3/16/24.
//

import EkycID
import Foundation
import AVFoundation

public class FlutterFaceScanner: NSObject, FlutterPlatformView, FaceScannerEventListener {
    let viewId: Int64
    var flutterScannerView: UIView?
    var scanner: FaceScannerView!
    var methodChannel: FlutterMethodChannel?
    var eventChannel: FlutterEventChannel?
    var eventStreamHandler: FaceScannerEventStreamHandler?
    
    init(frame: CGRect, viewId: Int64, messenger: FlutterBinaryMessenger, args: Any?) {
        let screenSize = UIScreen.main.bounds
        let screenWidth = screenSize.width
        let screenHeight = screenSize.height
        
        self.viewId = viewId
        
        super.init()
        
        self.flutterScannerView = UIView(frame: CGRect(x: 0, y: 0, width: screenWidth, height: screenHeight))
        self.methodChannel = FlutterMethodChannel(name: "FaceScanner_MethodChannel_" + String(viewId), binaryMessenger: messenger)
        self.eventChannel = FlutterEventChannel(name: "FaceScanner_EventChannel_" + String(viewId), binaryMessenger: messenger)
        self.eventStreamHandler = FaceScannerEventStreamHandler()
        self.methodChannel!.setMethodCallHandler(self.onMethodCall)
        self.eventChannel!.setStreamHandler(self.eventStreamHandler)
    }
    
    public func view() -> UIView {
        return self.flutterScannerView!
    }
    
    private func start(call: FlutterMethodCall, result: @escaping FlutterResult) throws {
        let args = call.arguments as! [String: Any?]
        let useFrontCamera = args["useFrontCamera"] as! Bool
        self.scanner = FaceScannerView(
            frame: self.flutterScannerView!.frame,
            useFrontCamera: useFrontCamera
        )
        self.scanner.addListener(self)
        self.scanner.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        self.flutterScannerView!.addSubview(self.scanner)
        self.scanner.start(options: self.argsToFaceScannerOptions(args))
        result(true)
    }
    
    private func argsToFaceScannerOptions(_ args: [String: Any?]) -> FaceScannerOptions {
        let cameraOptions = args["cameraOptions"] as! [String: Any?]
        let captureDurationCountDown = cameraOptions["captureDurationCountDown"] as! Int
        let faceCropScale = cameraOptions["faceCropScale"] as! NSNumber
        let roiSize = cameraOptions["roiSize"] as! NSNumber
        let minFaceWidthPercentage = cameraOptions["minFaceWidthPercentage"] as! NSNumber
        let maxFaceWidthPercentage = cameraOptions["maxFaceWidthPercentage"] as! NSNumber
        return FaceScannerOptions(
            cameraOptions: FaceScannerCameraOptions(
                captureDurationCountDown: captureDurationCountDown,
                faceCropScale: faceCropScale.floatValue,
                roiSize: roiSize.floatValue,
                minFaceWidthPercentage: minFaceWidthPercentage.floatValue,
                maxFaceWidthPercentage: maxFaceWidthPercentage.floatValue
            )
        )
    }
    
    public func onInitialized() {
        self.eventStreamHandler?.sendOnInitializedEventToFlutter()
    }
    
    public func onFaceScanned(_ face: EkycID.LivenessFace) {
        self.eventStreamHandler?.sendOnFaceScannedEventToFlutter(face)
    }
    
    public func onFrameStatusChanged(_ frameStatus: EkycID.FrameStatus) {
        self.eventStreamHandler?.sendOnFrameStatusChangedEventToFlutter(frameStatus)
    }
    
    public func onCaptureCountDownChanged(current: Int, max: Int) {
        self.eventStreamHandler?.sendOnCaptureCountDownChangedEventToFlutter(current: current, max: max)
    }
    
    private func nextImage(call: FlutterMethodCall, result: @escaping FlutterResult) throws {
        if self.scanner != nil {
            self.scanner.nextImage()
        }
        result(true)
    }
    
    private func dispose(call: FlutterMethodCall, result: @escaping FlutterResult) throws {
        if (self.scanner != nil) {
            self.scanner!.stop()
            self.scanner = nil
        }
        result(true)
    }
    
    private func onMethodCall(call: FlutterMethodCall, result: @escaping FlutterResult) -> Void {
        switch (call.method) {
        case "start":
            do {
                try self.start(call: call, result: result)
            } catch {
                result(FlutterError(code: "initialize Error", message: nil, details: nil))
            }
            break
        case "dispose":
            do {
                try self.dispose(call: call, result: result)
            } catch {
                result(FlutterError(code: "dispose Error", message: nil, details: nil))
            }
            break
        case "nextImage":
            do {
                try self.nextImage(call: call, result: result)
            } catch {
                result(FlutterError(code: "dispose Error", message: nil, details: nil))
            }
            break
        default:
            result(FlutterMethodNotImplemented)
            break
        }
    }
    
    class FaceScannerEventStreamHandler: NSObject, FlutterStreamHandler {
        var events: FlutterEventSink?
        
        func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
            self.events = events
            return nil
        }
        
        func onCancel(withArguments arguments: Any?) -> FlutterError? {
            return nil
        }
        
        func sendOnInitializedEventToFlutter() {
            if (self.events != nil) {
                DispatchQueue.main.async {
                    var event = [String: Any]()
                    event["type"] = "onInitialized"
                    self.events!(event)
                }
            }
        }
        
        func sendOnFrameStatusChangedEventToFlutter(_ frameStatus: FrameStatus) {
            if (self.events != nil) {
                DispatchQueue.main.async {
                    var event = [String: Any]()
                    event["type"] = "onFrameStatusChanged"
                    event["values"] = "\(frameStatus)"
                    self.events!(event)
                }
            }
        }
        
        func sendOnFaceScannedEventToFlutter(_ face: EkycID.LivenessFace) {
            if (self.events != nil) {
                DispatchQueue.main.async {
                    var event = [String: Any]()
                    event["type"] = "onFaceScanned"
                    event["values"] = face.toFlutterMap()
                    self.events!(event)
                }
            }
        }
        
        func sendOnCaptureCountDownChangedEventToFlutter(current: Int, max: Int) {
            if (self.events != nil) {
                DispatchQueue.main.async {
                    var event = [String: Any]()
                    event["type"] = "onCaptureCountDownChanged"
                    var result = [String: Any?]()
                    result["current"] = current
                    result["max"] = max
                    event["values"] = result
                    self.events!(event)
                }
            }
        }
    }
}

File: android/src/main/kotlin/com/ekycsolutions/ekyc_id_flutter/FaceScanner/FlutterFaceScanner.kt
package com.ekycsolutions.ekyc_id_flutter.FaceScanner

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.graphics.Bitmap
import android.view.View
import android.widget.LinearLayout
import com.ekycsolutions.ekyc_id_flutter.R
import com.ekycsolutions.ekycid.core.models.FrameStatus
import com.ekycsolutions.ekycid.facescanner.FaceScannerEventListener
import com.ekycsolutions.ekycid.facescanner.FaceScannerOptions
import com.ekycsolutions.ekycid.facescanner.FaceScannerView
import com.ekycsolutions.ekycid.facescanner.cameraview.FaceScannerCameraOptions
import com.ekycsolutions.ekycid.livenessdetection.cameraview.LivenessFace
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView
import java.io.ByteArrayOutputStream


class FlutterFaceScanner(
    private var binding: FlutterPlugin.FlutterPluginBinding,
    private var context: Context,
    private val viewId: Int
): PlatformView, MethodChannel.MethodCallHandler, FaceScannerEventListener {
    private var scanner: FaceScannerView? = null

    private var scannerView: LinearLayout = LinearLayout(context)
    private val methodChannel: MethodChannel =
        MethodChannel(binding.binaryMessenger, "FaceScanner_MethodChannel_$viewId")
    private val eventChannel: EventChannel =
        EventChannel(binding.binaryMessenger, "FaceScanner_EventChannel_$viewId")
    private val eventStreamHandler = FaceScannerEventStreamHandler(context)

    init {
        this.methodChannel.setMethodCallHandler(this)
        this.eventChannel.setStreamHandler(eventStreamHandler)
    }


    override fun getView(): View {
        return scannerView
    }

    override fun dispose() {

    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "start" -> {
                start(call, result)
            }
            "nextImage" -> {
                nextImage(call, result)
            }
            "dispose" -> {
                disposeFlutter(call, result)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    private fun start(call: MethodCall, result: MethodChannel.Result) {
        try {
            val args = call.arguments as HashMap<*, *>
            val useFrontCamera = args["useFrontCamera"] as Boolean
            val cameraOptions = args["cameraOptions"] as HashMap<String, Any>
            this.scanner = FaceScannerView(context, useFrontCamera)
            this.scanner!!.addListener(this)
            this.scanner!!.start(
                FaceScannerOptions(
                    FaceScannerCameraOptions(
                        cameraOptions["captureDurationCountDown"] as Int,
                        cameraOptions["faceCropScale"] as Float,
                        cameraOptions["roiSize"] as Float,
                        cameraOptions["minFaceWidthPercentage"] as Float,
                        cameraOptions["maxFaceWidthPercentage"] as Float,
                    )
                )
            )

            this.scanner!!.layoutParams = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.MATCH_PARENT
            )
            this.scannerView.addView(this.scanner)
            result.success(true)
        } catch (e: Exception) {
            result.error(e.toString(), e.message, "")
        }
    }

    private fun disposeFlutter(call: MethodCall, result: MethodChannel.Result) {
        try {
            if (this.scanner != null) {
                this.scanner!!.stop()
                this.scanner = null
            }
            result.success(true)
        } catch (e: Exception) {
            result.error(e.toString(), e.message, "")
        }
    }

    private fun nextImage(call: MethodCall, result: MethodChannel.Result) {
        try {
            if (this.scanner != null) {
                this.scanner!!.nextImage()
            }
            result.success(true)
        } catch (e: Exception) {
            result.error(e.toString(), e.message, "")
        }
    }

    override fun onInitialized() {
        this.eventStreamHandler?.sendOnInitializedEventToFlutter()
    }

    override fun onFaceScanned(face: LivenessFace) {
        this.eventStreamHandler?.sendOnFaceScannedEventToFlutter(face)
    }

    override fun onFrameStatusChanged(frameStatus: FrameStatus) {
        this.eventStreamHandler?.sendOnFrameStatusChangedEventToFlutter(frameStatus)
    }

    override fun onCaptureCountDownChanged(current: Int, max: Int) {
        this.eventStreamHandler?.sendOnCaptureCountDownChangedEventToFlutter(current, max)
    }

    class FaceScannerEventStreamHandler(private var context: Context) : EventChannel.StreamHandler {
        private var events: EventChannel.EventSink? = null

        override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
            this.events = events
        }

        override fun onCancel(arguments: Any?) {
            this.events = null
        }

        fun sendOnInitializedEventToFlutter() {
            if (events != null) {
                (context as Activity).runOnUiThread {
                    val event = HashMap<String, Any>()
                    event["type"] = "onInitialized"
                    val values = HashMap<String, Any?>()
                    event["values"] = values
                    events!!.success(event)
                }
            }
        }

        fun sendOnFrameStatusChangedEventToFlutter(frameStatus: FrameStatus) {
            if (events != null) {
                (context as Activity).runOnUiThread {
                    val event = HashMap<String, Any>()
                    event["type"] = "onFrameStatusChanged"
                    event["values"] = frameStatus.name
                    events!!.success(event)
                }
            }
        }

        fun sendOnFaceScannedEventToFlutter(face: LivenessFace) {
            if (events != null) {
                (context as Activity).runOnUiThread {
                    val event = HashMap<String, Any>()
                    event["type"] = "onFaceScanned"
                    event["values"] = livenessFaceToFlutterMap(face)
                    events!!.success(event)
                }
            }
        }

        fun sendOnCaptureCountDownChangedEventToFlutter(current: Int, max: Int) {
            if (events != null) {
                (context as Activity).runOnUiThread {
                    val event = HashMap<String, Any>()
                    event["type"] = "onCaptureCountDownChanged"
                    val values = HashMap<String, Any?>()
                    values["current"] = current
                    values["max"] = max
                    event["values"] = values
                    events!!.success(event)
                }
            }
        }

        private fun livenessFaceToFlutterMap(livenessFace: LivenessFace): HashMap<String, Any?> {
            var values = HashMap<String, Any?>()
            values["image"] = bitmapToFlutterByteArray(livenessFace.image!!)
            values["leftEyeOpenProbability"] = livenessFace.leftEyeOpenProbability
            values["rightEyeOpenProbability"] = livenessFace.rightEyeOpenProbability
            values["headEulerAngleX"] = livenessFace.headEulerAngleX
            values["headEulerAngleY"] = livenessFace.headEulerAngleY
            values["headEulerAngleZ"] = livenessFace.headEulerAngleZ

            if (livenessFace.headDirection != null) {
                values["headDirection"] = livenessFace.headDirection!!.name
            } else {
                values["headDirection"] = null
            }

            if (livenessFace.eyesStatus != null) {
                values["eyesStatus"] = livenessFace.eyesStatus!!.name
            } else {
                values["eyesStatus"] = null
            }

            return values
        }

        private fun bitmapToFlutterByteArray(image: Bitmap): ByteArray {
            val stream = ByteArrayOutputStream()
            image.compress(Bitmap.CompressFormat.JPEG, 90, stream)
            return stream.toByteArray()
        }
    }
}